editor/components/inserter/menu.js

```diff
@@ -29,32 +77,206 @@ class InserterMenu extends wp.element.Component {
-			return groups;
+
+			accumulator[ block.category ] = [ block ];
+			return accumulator;
 		}, {} );
-		const categories = wp.blocks.getCategories();
+	}
+
+	getVisibleBlocksByCategory( blockTypes ) {
+		return flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+			this.groupByCategory
+		)( blockTypes );
+	}
+
+	findNext( currentBlock, blockTypes ) {
+		/**
+		 * null is the value that will trigger iterating back to
+		 * the top of the list of block types.
+		 */
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
+		const nextIndex = currentIndex + 1;
+		const highestIndex = blockTypes.length - 1;
+
+		/**
+		 * Default currently for going past the blocks is search, may need to be
+		 * revised in the future as more focusable elements are added. This
+		 * returns a null value, which currently implies that search will be set
+		 * as the next focus.
+		 */
+		if ( nextIndex > highestIndex ) {
+			return null;
+		}
+
+		// Return the slug of the next block type.
+		return blockTypes[ nextIndex ].slug;
+	}
+
+	findPrevious( currentBlock, blockTypes ) {
+		/**
+		 * null will trigger iterating back to the top of the list of block
+		 * types.
+		 */
+		if ( null === currentBlock ) {
+			return blockTypes[ 0 ].slug;
+		}
+
+		const highestIndex = blockTypes.length - 1;
+
+		// If the search bar is focused navigate to the bottom of the block list.
+		if ( 'search' === currentBlock ) {
+			return blockTypes[ highestIndex ].slug;
+		}
+
+		const currentIndex = blockTypes.findIndex( ( blockType ) => currentBlock === blockType.slug );
+		const previousIndex = currentIndex - 1;
+		const lowestIndex = 0;
+
+		/**
+		 * Default currently for going past the blocks is search, may need to be
+		 * revised in the future as more focusable elements are added. This
+		 * returns a null value, which currently implies that search will be set
+		 * as the next focus.
+		 */
+		if ( previousIndex < lowestIndex ) {
+			return null;
+		}
+
+		// Return the slug of the next block type.
+		return blockTypes[ previousIndex ].slug;
+	}
+
+	focusNext( component ) {
+		const sortedByCategory = flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+		)( component.blockTypes );
+
+		// If the block list is empty return early.
+		if ( ! sortedByCategory.length ) {
+			return;
+		}
+
+		const currentBlock = component.state.currentFocus;
+
+		const nextBlock = this.findNext( currentBlock, sortedByCategory );
+		this.changeMenuSelection( nextBlock );
+	}
+
+	focusPrevious( component ) {
+		const sortedByCategory = flow(
+			this.getVisibleBlocks,
+			this.sortBlocksByCategory,
+		)( component.blockTypes );
+		const currentBlock = component.state.currentFocus;
+
+		// If the block list is empty return early.
+		if ( ! sortedByCategory.length ) {
+			return;
+		}
+
+		const nextBlock = this.findPrevious( currentBlock, sortedByCategory );
```

Comment:
We should treat `state` as a single source of truth; since `findPrevious` has access to `this.state`, there's no need to pass it as an argument.